File: planes/moose/app/apis/aircraftSpeedAltitudeByType.ts
import { Api, getMooseUtils, WebApp } from "@514labs/moose-lib";
import { AircraftTrackingProcessed_Table } from "../index";
import express, { Request } from "express";
import cors from "cors";

/**
 * Parameters for the aircraft speed and altitude by type API
 */
interface AircraftSpeedAltitudeParams {
  /** Optional filter for specific aircraft category */
  category?: string;
  /** Optional minimum barometric altitude filter */
  minAltitude?: number;
  /** Optional maximum barometric altitude filter */
  maxAltitude?: number;
  /** Optional minimum ground speed filter */
  minSpeed?: number;
  /** Optional maximum ground speed filter */
  maxSpeed?: number;
}

/**
 * Constructs the SQL query for aircraft speed and altitude statistics
 */
const buildAircraftStatsQuery = (sql: any, aircraft_cols: any, params: AircraftSpeedAltitudeParams) => {
  return sql`
    SELECT
      ${aircraft_cols.category} as aircraft_category,
      COUNT(*) as total_records,
      AVG(${aircraft_cols.alt_baro}) as avg_barometric_altitude,
      MIN(${aircraft_cols.alt_baro}) as min_barometric_altitude,
      MAX(${aircraft_cols.alt_baro}) as max_barometric_altitude,
      STDDEV_POP(${aircraft_cols.alt_baro}) as altitude_stddev,
      AVG(${aircraft_cols.gs}) as avg_ground_speed,
      MIN(${aircraft_cols.gs}) as min_ground_speed,
      MAX(${aircraft_cols.gs}) as max_ground_speed,
      STDDEV_POP(${aircraft_cols.gs}) as speed_stddev,
      COUNT(DISTINCT ${aircraft_cols.hex}) as unique_aircraft_count
    FROM ${AircraftTrackingProcessed_Table}
    WHERE ${aircraft_cols.alt_baro} > 0
      AND ${aircraft_cols.gs} > 0
      AND ${aircraft_cols.category} != ''
      -- Optional category filter
      AND (${params.category || ''} = '' OR ${aircraft_cols.category} = ${params.category || ''})
      -- Optional altitude range filters
      AND (${params.minAltitude || -999999} = -999999 OR ${aircraft_cols.alt_baro} >= ${params.minAltitude || -999999})
      AND (${params.maxAltitude || 999999} = 999999 OR ${aircraft_cols.alt_baro} <= ${params.maxAltitude || 999999})
      -- Optional speed range filters
      AND (${params.minSpeed || -999999} = -999999 OR ${aircraft_cols.gs} >= ${params.minSpeed || -999999})
      AND (${params.maxSpeed || 999999} = 999999 OR ${aircraft_cols.gs} <= ${params.maxSpeed || 999999})
    GROUP BY ${aircraft_cols.category}
    ORDER BY total_records DESC
  `;
};


const app = express();
app.use(cors());
app.use(express.json());

/**
 * Express API Handler
 * API that provides speed and altitude statistics for different aircraft types/categories
 * Uses barometric altitude (NOT geometric altitude) and ground speed
 */
app.get("/aircraftSpeedAltitudeByType", async (req: Request<{}, {}, {}, AircraftSpeedAltitudeParams>, res) => {
  const { client, sql } = await getMooseUtils();

  // Reference the source table object inside the function
  const aircraft_cols = AircraftTrackingProcessed_Table?.columns;
  const params = req.query;

  // Execute the query with robust optional parameter handling
  const query = buildAircraftStatsQuery(sql, aircraft_cols, params);
  const result = await client.query.execute(query);

  const data = await result.json();
  res.json(data);
});

new WebApp("aircraft", app, {
  mountPath: "/aircraft/api",
  metadata: {
    description:
      "API for consuming aircraft data",
  },
});

/**
 * Moose Consumption Api
 * API that provides speed and altitude statistics for different aircraft types/categories
 * Uses barometric altitude (NOT geometric altitude) and ground speed

export const aircraftSpeedAltitudeByType = new Api<AircraftSpeedAltitudeParams>(
  "aircraftSpeedAltitudeByType",
  async (params, { client, sql }) => {
    // Reference the source table object inside the function
    const aircraft_cols = AircraftTrackingProcessed_Table?.columns;

    // Execute the query with robust optional parameter handling
    const query = buildAircraftStatsQuery(sql, aircraft_cols, params);
    const result = await client.query.execute(query);

    return result;
  },
  {
    metadata: {
      description: "Provides speed and altitude statistics for different aircraft types/categories",
    },
  }
);
 */

File: planes/moose-py/app/apis/aircraft_speed_altitude_by_type.py
from typing import Optional, Any, List
import logging

from fastapi import FastAPI, Query, Request, HTTPException  # pyright: ignore[reportMissingImports]
from fastapi.middleware.cors import CORSMiddleware  # pyright: ignore[reportMissingImports]
from moose_lib.dmv2 import WebApp, WebAppConfig, WebAppMetadata
from moose_lib.dmv2.web_app_helpers import get_moose_utils
from pydantic import BaseModel

logger = logging.getLogger(__name__)

from app.ingest.aircraft_tracking import AircraftTrackingProcessed_Table

app = FastAPI(
    title="Aircraft API",
    description="API for consuming aircraft data",
    version="1.0.0",
)

# Keep this permissive to match the TS Express implementation (which enables CORS).
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


class AircraftSpeedAltitudeByTypeRow(BaseModel):
    aircraft_category: str
    total_records: int
    avg_barometric_altitude: float
    min_barometric_altitude: float
    max_barometric_altitude: float
    altitude_stddev: float
    avg_ground_speed: float
    min_ground_speed: float
    max_ground_speed: float
    speed_stddev: float
    unique_aircraft_count: int


@app.get("/aircraftSpeedAltitudeByType")
async def aircraftSpeedAltitudeByType(
    request: Request,
    category: Optional[str] = Query(default=None),
    minAltitude: Optional[float] = Query(default=None),
    maxAltitude: Optional[float] = Query(default=None),
    minSpeed: Optional[float] = Query(default=None),
    maxSpeed: Optional[float] = Query(default=None),
):
    """
    Provides speed and altitude statistics for different aircraft types/categories.
    Uses barometric altitude (NOT geometric altitude) and ground speed.
    """
    try:
        moose = get_moose_utils(request)
        if not moose:
            raise HTTPException(status_code=500, detail="MooseStack utilities not available")
        
        source = AircraftTrackingProcessed_Table
        
        query = f"""
            SELECT 
                {source.columns.category} as aircraft_category,
                COUNT(*) as total_records,
                AVG({source.columns.alt_baro}) as avg_barometric_altitude,
                MIN({source.columns.alt_baro}) as min_barometric_altitude,
                MAX({source.columns.alt_baro}) as max_barometric_altitude,
                STDDEV_POP({source.columns.alt_baro}) as altitude_stddev,
                AVG({source.columns.gs}) as avg_ground_speed,
                MIN({source.columns.gs}) as min_ground_speed,
                MAX({source.columns.gs}) as max_ground_speed,
                STDDEV_POP({source.columns.gs}) as speed_stddev,
                COUNT(DISTINCT {source.columns.hex}) as unique_aircraft_count
            FROM {source.name} 
            WHERE {source.columns.alt_baro} > 0 
                AND {source.columns.gs} > 0 
                AND {source.columns.category} != ''
                -- Optional category filter
                AND ('{category or ''}' = '' OR {source.columns.category} = '{category or ''}')
                -- Optional altitude range filters
                AND ({minAltitude or -999999} = -999999 OR {source.columns.alt_baro} >= {minAltitude or -999999})
                AND ({maxAltitude or 999999} = 999999 OR {source.columns.alt_baro} <= {maxAltitude or 999999})
                -- Optional speed range filters
                AND ({minSpeed or -999999} = -999999 OR {source.columns.gs} >= {minSpeed or -999999})
                AND ({maxSpeed or 999999} = 999999 OR {source.columns.gs} <= {maxSpeed or 999999})
            GROUP BY {source.columns.category} 
            ORDER BY total_records DESC
        """

        # Execute query and get results
        result = moose.client.query.execute_raw(query, parameters={})

        # Handle result - some implementations return a wrapper with .json() method
        if hasattr(result, "json") and callable(getattr(result, "json")):
            return result.json()

        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Error in aircraftSpeedAltitudeByType")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


# Register as MooseStack WebApp (FastAPI mounted into the project)
aircraft_fastapi_app = WebApp(
    "aircraft",
    app,
    WebAppConfig(
        mount_path="/aircraft/api",
        metadata=WebAppMetadata(description="API for consuming aircraft data"),
    ),
)

